File: app/src/main/java/edu/upc/mcia/publications/data/repository/AuthorRepository.java
package edu.upc.mcia.publications.data.repository;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import edu.upc.mcia.publications.data.model.Author;
import edu.upc.mcia.publications.data.remote.ApiManager;
import io.reactivex.Observable;

public class AuthorRepository {
    private static AuthorRepository INSTANCE = null;

    private final Map<String, Author> mData = new HashMap<>();

    private AuthorRepository() {
    }

    public static AuthorRepository getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new AuthorRepository();
        }
        return INSTANCE;
    }

    public Observable<List<Author>> findAll() {
        if (mData.isEmpty()) {
            return ApiManager.getRemoteApi().getAuthors()
                    .concatMap(Observable::fromIterable)
                    .doOnNext(a -> mData.put(a.getId(), a))
                    .toList().toObservable();
        } else {
            return Observable.fromIterable(mData.values()).toList().toObservable();
        }
    }

    public Observable<Author> findById(String id) {
        if (!mData.containsKey(id)) {
            return ApiManager.getRemoteApi().getAuthor(id)
                    .doOnNext(a -> mData.put(a.getId(), a));
        } else {
            return Observable.just(mData.get(id));
        }
    }

}

File: app/src/main/java/edu/upc/mcia/publications/data/repository/AuthorRepository.kt
package edu.upc.mcia.publications.data.repository

import edu.upc.mcia.publications.data.model.Author
import edu.upc.mcia.publications.data.remote.ApiManager
import io.reactivex.Observable
import java.util.*

object AuthorRepository {

    private val mData = HashMap<String, Author>()

    fun findAll(): Observable<List<Author>> {
        return if (mData.isEmpty()) {
            ApiManager.remoteApi.authors
                    .concatMap<Author>({ Observable.fromIterable(it) })
                    .doOnNext { a -> mData.put(a.id, a) }
                    .toList().toObservable()
        } else {
            Observable.fromIterable(mData.values).toList().toObservable()
        }
    }

    fun findById(id: String): Observable<Author> {
        return if (!mData.containsKey(id)) {
            ApiManager.remoteApi.getAuthor(id)
                    .doOnNext { a -> mData.put(a.id, a) }
        } else {
            Observable.just(mData[id])
        }
    }

}
